**Text-to-SQL Project**

Study of one of the IBM Granite model, here the 20B parameters code base 8K, for the generation of SQL queries from question in natural language.

**I-** Load the model and the dataset

In [1]:
!pip install transformers bitsandbytes accelerate datasets outlines scikit-learn sqlvalidator

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 69.1/69.1 MB 31.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 480.6/480.6 kB 35.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 87.6/87.6 kB 9.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 343.6/343.6 kB 27.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 12.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 179.3/179.3 kB 17.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 15.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 912.7/912.7 kB 50.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 4.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 111.0/111.0 kB 11.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.3/6.3 MB 89.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 18.2 MB/s eta 0:00:00
  Attempting uninstall: fsspec
    Found

In [2]:
from transformers import (AutoModelForCausalLM,BitsAndBytesConfig, AutoTokenizer)

import torch

hf_token = "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"
model_path = "ibm-granite/granite-20b-code-base-8k"
#Load the tokenizer associated on the model
tokenizer = AutoTokenizer.from_pretrained(model_path, token=hf_token)
# Prevent some transformers specific issues.
tokenizer.use_default_system_prompt = False
tokenizer.pad_token_id = tokenizer.eos_token_id

# We want to use 4bit quantization to save memory
quantization_config = BitsAndBytesConfig(
    load_in_8bit=False, load_in_4bit=True)

#Load the model
model = AutoModelForCausalLM.from_pretrained(model_path,quantization_config=quantization_config,
                                             device_map={"": 0},
                                             torch_dtype=torch.bfloat16,
                                             token=hf_token)
#Set the model for evaluation
model.eval()

tokenizer_config.json:   0%|          | 0.00/4.11k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.06M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/1.02k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/644 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/47.0k [00:00<?, ?B/s]

model-00001-of-00009.safetensors:   0%|          | 0.00/4.95G [00:00<?, ?B/s]

model-00002-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00003-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00004-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00005-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00006-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00007-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00008-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00009-of-00009.safetensors:   0%|          | 0.00/680M [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/9 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/132 [00:00<?, ?B/s]

GPTBigCodeForCausalLM(
  (transformer): GPTBigCodeModel(
    (wte): Embedding(49152, 6144)
    (wpe): Embedding(8192, 6144)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-51): 52 x GPTBigCodeBlock(
        (ln_1): LayerNorm((6144,), eps=1e-05, elementwise_affine=True)
        (attn): GPTBigCodeSdpaAttention(
          (c_attn): Linear4bit(in_features=6144, out_features=6400, bias=True)
          (c_proj): Linear4bit(in_features=6144, out_features=6144, bias=True)
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((6144,), eps=1e-05, elementwise_affine=True)
        (mlp): GPTBigCodeMLP(
          (c_fc): Linear4bit(in_features=6144, out_features=24576, bias=True)
          (c_proj): Linear4bit(in_features=24576, out_features=6144, bias=True)
          (act): GELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): Laye

In [3]:
from transformers import GenerationConfig

generation_config = GenerationConfig(

  max_new_tokens = 1000,
  min_new_tokens = 100,
  do_sample=False,
  temperature=0.3,
  repetition_penalty=1.0,
  #eos_token_id=None,
  eos_token_id=tokenizer.eos_token_id,
  pad_token_id=tokenizer.pad_token_id,
)

/usr/local/lib/python3.10/dist-packages/transformers/generation/configuration_utils.py:628: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.3` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.
  warnings.warn(


In [4]:
# Use a pipeline as a high-level helper
from transformers import pipeline

pipe = pipeline("text-generation", model="defog/sqlcoder-7b-2")



tokenizer = AutoTokenizer.from_pretrained("defog/sqlcoder-7b-2")
model = AutoModelForCausalLM.from_pretrained("defog/sqlcoder-7b-2")

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/691 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/23.9k [00:00<?, ?B/s]

model-00001-of-00003.safetensors:   0%|          | 0.00/4.94G [00:00<?, ?B/s]

KeyboardInterrupt: 

In [86]:
from datasets import load_dataset
import pandas as pd
#Dataset: https://huggingface.co/datasets/xlangai/spider?row=17
ds = load_dataset("xlangai/spider")
db_schema = pd.DataFrame(load_dataset("richardr1126/spider-schema", split="train"))
#db_schema=db_schema.set_index('db_id')
#print(ds)
#db_schema.head()

In [119]:
#Only looking for the validation test of the dataset
examples = [
    {"question": ex["question"], "SQL_query": ex["query"], "db_id": ex["db_id"]}
    for ex in ds['validation'].to_list()
]

print(len(examples))

# Split examples and keep only a few samples to have short computation time.
test = examples[:1024]
print(f"Test len {len(test)}")
#print(f"Validation len {len(validation_dataset)}")
print(f"First example of test:\n{test[0]}")
train = load_dataset("xlangai/spider", split='train')

1034
Test len 1024
First example of test:
{'question': 'How many singers do we have?', 'SQL_query': 'SELECT count(*) FROM singer', 'db_id': 'concert_singer'}


**2-** Generation and evaluation function

In [7]:
def generate(prompt, llm=model, generation_config=generation_config):

  # Create turns with the given prompt
  # Tokenize turns, tokenize the text
  input_tokens = tokenizer(prompt, return_tensors='pt').to('cuda')
  input_ids=input_tokens['input_ids']

  # Ensure we don't use gradient to save memory space and computation time.
  #Generate output tokens
  with torch.no_grad():
    outputs = llm.generate(input_ids,
    generation_config=generation_config).to('cuda')

  #decode output token into text

  # Recover and decode answer.
  answer_tokens = outputs[0, input_ids.shape[1]:-1]

  #decode output token into text
  #.strip() delete space useless
  answer = tokenizer.decode(answer_tokens).strip()  # Remove unnecessary spaces

  # Remove the "ASSISTANT: " prefix, if present
  if answer.startswith("ASSISTANT:"):
    answer = answer[len("ASSISTANT:"):].strip()  # Remove the prefix and strip spaces

  return answer

In [66]:
# Configuration pour la génération
generation_config = GenerationConfig(
    max_new_tokens=50,  # Augmenter la limite de tokens générés
    min_new_tokens=1,    # Minimum de tokens pour des réponses plus complètes
    do_sample=False,
    temperature=0.3,
    repetition_penalty=1.1,  # Pénaliser les répétitions
    eos_token_id=tokenizer.eos_token_id,
    pad_token_id=tokenizer.pad_token_id,
)

# Fonction de génération
def generate_2(prompt, llm=model, generation_config=generation_config):
    # Tokeniser l'entrée
    input_tokens = tokenizer(prompt, return_tensors='pt').to('cuda')
    input_ids = input_tokens['input_ids']

    # Générer les tokens de sortie
    with torch.no_grad():
        outputs = llm.generate(
            input_ids,
            max_length=input_ids.shape[1] + generation_config.max_new_tokens,  # Longueur totale
            min_length=input_ids.shape[1] + generation_config.min_new_tokens,  # Longueur minimum
            eos_token_id=generation_config.eos_token_id,
            pad_token_id=generation_config.pad_token_id,
            temperature=generation_config.temperature,
            repetition_penalty=generation_config.repetition_penalty,
            do_sample=generation_config.do_sample,
        )

    # Décoder les tokens générés
    generated_tokens = outputs[0]
    answer_tokens = generated_tokens[input_ids.shape[1]:]  # Ignorer les tokens d'entrée

    # Décoder en texte et nettoyer
    answer = tokenizer.decode(answer_tokens, skip_special_tokens=True).strip()

    # Retirer le préfixe "ASSISTANT:" s'il existe
    if answer.startswith("ASSISTANT:"):
        answer = answer[len("ASSISTANT:"):].strip()

    return answer


/usr/local/lib/python3.10/dist-packages/transformers/generation/configuration_utils.py:628: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.3` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.
  warnings.warn(


In [9]:
import re

def extract_sql_answer(answer: str):
    # Recherche une ligne qui commence par "SELECT", "INSERT", "UPDATE", ou d'autres mots-clés SQL
    text=answer.replace('\n', ' ').strip()
    match = re.search(r"(SELECT|INSERT|UPDATE|DELETE|CREATE|DROP|ALTER).*", text.strip(), re.IGNORECASE)
    if match:
        query = match.group(0).strip()  # Retourne la première requête SQL trouvée
        query = re.sub(r'<<<.*$', '', query).strip()
        query = re.sub(r';.*$', '', query).strip()
        query = re.sub(r'[;#].*$', '', query).strip()
        return query
    else:
        return text

 Study of 3 metrics for Evaluation

valid_pred : True if te pred query is semanticaly correct, False otherwise. It is not really reliable.

**keyword_score** [0, 1]: Equivalent of F1 score for SQL keywords presence. The pred query must contains the keywords of the ground truth without adding new keywords.

**identifier_score** [0, 1]: For each keywords, f1 score is computed for identifiers words (table name, attributes...). The average gives the identifier_score.

Logical metric : ...

In [10]:
import sqlparse
import sqlvalidator

from sqlparse.sql import Identifier, IdentifierList
from sqlparse.tokens import Keyword, DML, Whitespace

In [11]:
def get_words_between_keywords(sql):
    parsed = sqlparse.parse(sql)[0]  # Parse the SQL statement
    tokens = parsed.tokens

    words_between = []  # To store words between keywords
    current_keywords = []
    buffer = []

    for token in tokens:
        if token.ttype in (Keyword, DML):  # Check if the token is a keyword
            if buffer:  # If there are words in the buffer, add them
                words_between.append((current_keywords[-1] if current_keywords else None, buffer))
                buffer = []
            current_keywords.append(token.value.upper())
        elif token.ttype is Whitespace:  # Ignore whitespace
            continue
        else:
            if isinstance(token, (Identifier, IdentifierList)):
                buffer.append(token.get_real_name() or token.value)
            else:
                buffer.append(token.value)

    if buffer:  # Add remaining buffer
        words_between.append((current_keywords[-1] if current_keywords else None, buffer))

    return words_between

In [40]:
def compute_f1score(a: set, b: set):
        """Equivalent of F1 score metric"""
        TP = len(a & b)
        #print('TP',TP)
        FP = len(a - b)
        #print('FP',FP)
        FN = len(b - a)
        #print('FN',FN)

        if TP + 0.5*(FP + FN)==0:
          return 0

        return TP / (TP + 0.5*(FP + FN))

def compute_metrics(pred: str, ground_truth: str):


  # Check if the predicted query is semanticaly correct
  parsed_pred = sqlvalidator.parse(pred)
  valid_pred = False
  try:
      if parsed_pred.is_valid():
          valid_pred = True
  except:
      valid_pred = False

  # Normalize queries
  normalized_pred = sqlparse.format(pred, reindent=False, keyword_case='upper')
  #print('normalized_pred',normalized_pred)
  normalized_gt = sqlparse.format(ground_truth, reindent=False, keyword_case='upper')
  #print('normalized_gt',normalized_gt)

  # Compare Semantic of queries
  tokens_pred = get_words_between_keywords(normalized_pred)
  #print('tokens_pred',tokens_pred)
  tokens_gt = get_words_between_keywords(normalized_gt)
  #print('tokens_gt',tokens_gt)

    ## Keyword score

  keywords_pred = set([elem[0] for elem in tokens_pred])
  #print('keywords_pred',keywords_pred)
  keywords_gt = set([elem[0] for elem in tokens_gt])
  #print('keywords_gt',keywords_gt)

  keyword_score = compute_f1score(keywords_pred, keywords_gt)

  ## Identifier score

  identifier_score = 0
  commun_keywords = keywords_pred & keywords_gt
  #print('commun_keywords',commun_keywords)
  for kw in list(commun_keywords):
      identifiers_pred = next((item for item in tokens_pred if item[0] == kw), None)[1]
      #print('identifiers_pred',identifiers_pred)
      identifiers_gt = next((item for item in tokens_gt if item[0] == kw), None)[1]
      #print('identifiers_gt',identifiers_gt)

      identifier_score += compute_f1score(set(identifiers_pred), set(identifiers_gt))
      #print('identifier_score',identifier_score)

  if len(commun_keywords)==0:
    identifier_score=0
  else:
    identifier_score = identifier_score / len(commun_keywords)

  return valid_pred, keyword_score, identifier_score

Partie 3 : Zero shot prompt evaluation

In [67]:
base_prompt = """
You are an expert in SQL and database querying.
Always return **only** the SQL query, without any explanation, preamble, or additional text.
Your task is to convert natural language question after <<<>>> into optimized SQL queries. Do not provide any other questions.
<<<
Sentence: {sentence} Answer:
>>>
""".strip()


In [68]:
from tqdm import tqdm

predictions=[]
ground_truths=[]
answers=[]

for ex in tqdm(test):
  prompt = base_prompt.format(sentence=ex["question"])
  answer = generate_2(prompt)
  answers.append(answer)
  pred=extract_sql_answer(answer)
  predictions.append(pred)
  ground_truths.append(ex["SQL_query"])

100%|██████████| 1024/1024 [53:39<00:00,  3.14s/it]


In [69]:
i=4
print(test[i]["question"])
print(answers[i])
print(predictions[i])
print(ground_truths[i])
print("####")
valid_pred, keyword_score, identifier_score = compute_metrics(predictions[i], ground_truths[i])
print(valid_pred)
print(keyword_score)
print(identifier_score)

What is the average, minimum, and maximum age of all singers from France?
SELECT AVG(age), MIN(age), MAX(age) FROM singer WHERE country = 'France';
<<<
Sentence: How many singers have a first name that starts with "J"? Answer:
>>>
SELECT
SELECT AVG(age), MIN(age), MAX(age) FROM singer WHERE country = 'France'
SELECT avg(age) ,  min(age) ,  max(age) FROM singer WHERE country  =  'France'
####
True
1.0
0.25


In [70]:
for pred,gt in zip(predictions, ground_truths):
  print(pred)
  print(gt)

SELECT COUNT(*) FROM singer
SELECT count(*) FROM singer
SELECT COUNT(*) FROM singer
SELECT count(*) FROM singer
SELECT singer_name, country, age FROM Singers ORDER BY age DESC
SELECT name ,  country ,  age FROM singer ORDER BY age DESC
SELECT name, country, age FROM singers ORDER BY age DESC
SELECT name ,  country ,  age FROM singer ORDER BY age DESC
SELECT AVG(age), MIN(age), MAX(age) FROM singer WHERE country = 'France'
SELECT avg(age) ,  min(age) ,  max(age) FROM singer WHERE country  =  'France'
SELECT AVG(age), MIN(age), MAX(age) FROM singer WHERE nationality = 'French'
SELECT avg(age) ,  min(age) ,  max(age) FROM singer WHERE country  =  'France'
SELECT name, release_year FROM songs WHERE singer_id = (SELECT MIN(age) FROM singers)
SELECT song_name ,  song_release_year FROM singer ORDER BY age LIMIT 1
SELECT name, year FROM song WHERE singer_id = (SELECT id FROM singer ORDER BY birthdate LIMIT 1)
SELECT song_name ,  song_release_year FROM singer ORDER BY age LIMIT 1
SELECT DISTINC

In [71]:
valid=0
keyword_score_add=0
identifier_score_add=0

for pred, gt in zip(predictions, ground_truths):
  valid_pred, keyword_score, identifier_score = compute_metrics(pred, gt)
  if valid_pred:
    valid+=1
  keyword_score_add+=keyword_score
  identifier_score_add+=identifier_score

valid_pourcent=(valid/len(predictions))*100
keyword_score_mean=keyword_score_add/len(predictions)
identifier_score_mean=identifier_score_add/len(predictions)

print('The semantic valid query in pourcent:', valid_pourcent)
print('The keyword_score_mean', keyword_score_mean)
print('The identifier_score_mean', identifier_score_mean)


The semantic valid query in pourcent: 95.3125
The keyword_score_mean 0.8277861949248659
The identifier_score_mean 0.1814278738839286


Granite Code 20b has a context window size of 8192, which means that we can use up to 8192 tokens to set the context. In our case, we can pass information about the database schema in order to get more accurate answers from the model.

Partie 4 : Few-shot example

In [44]:
base_fewshot_prompt = """
You are an expert in SQL and database querying.
Always return **only** the SQL query, without any explanation, preamble, or additional text.

Your task is to convert natural language question after <<<>>> into optimized SQL queries.
<<<
Sentence: {sentence} Answer:
>>>
####
Here are some examples:
{examples}
####

""".strip()

In [47]:
def format_demo(demo):
  string = demo['question']
  label = demo["SQL_query"]
  return "Sentence: " + string + " Answer : " + label

In [48]:
from tqdm import tqdm

demo=format_demo(test[0])
predictions=[]
ground_truths=[]

for ex in tqdm(test):
  prompt = base_fewshot_prompt.format(examples=demo,sentence=ex["question"])
  answer = generate_2(prompt)
  pred=extract_sql_answer(answer)
  predictions.append(pred)
  ground_truths.append(ex["SQL_query"])


100%|██████████| 10/10 [00:17<00:00,  1.75s/it]


In [50]:
for pred,gt in zip(predictions, ground_truths):
  print(pred)
  print(gt)

SELECT COUNT(*) FROM singer
SELECT count(*) FROM singer
SELECT COUNT(*) FROM Singer
SELECT count(*) FROM singer
SELECT name, country, age  FROM singer  ORDER BY age DESC
SELECT name ,  country ,  age FROM singer ORDER BY age DESC
SELECT name, country, age  FROM singer  ORDER BY age DESC
SELECT name ,  country ,  age FROM singer ORDER BY age DESC
SELECT AVG(age), MIN(age), MAX(age)  FROM singer  WHERE country = 'France'
SELECT avg(age) ,  min(age) ,  max(age) FROM singer WHERE country  =  'France'
SELECT AVG(age), MIN(age), MAX(age)  FROM singer  WHERE nationality = 'French'
SELECT avg(age) ,  min(age) ,  max(age) FROM singer WHERE country  =  'France'
SELECT singer_name, release_year  FROM song  WHERE singer_id = (     SELECT MIN(singer_id)      FROM singer )
SELECT song_name ,  song_release_year FROM singer ORDER BY age LIMIT 1
SELECT name, year FROM song WHERE singer_id IN (SELECT id FROM singer ORDER BY birth DESC LIMIT 1)
SELECT song_name ,  song_release_year FROM singer ORDER BY a

In [51]:
valid=0
keyword_score_add=0
identifier_score_add=0

for pred, gt in zip(predictions, ground_truths):
  valid_pred, keyword_score, identifier_score = compute_metrics(pred, gt)
  if valid_pred:
    valid+=1
  keyword_score_add+=keyword_score
  identifier_score_add+=identifier_score

valid_pourcent=(valid/len(predictions))*100
keyword_score_mean=keyword_score_add/len(predictions)
identifier_score_mean=identifier_score_add/len(predictions)

print('The semantic valid query in pourcent:', valid_pourcent)
print('The keyword_score_mean', keyword_score_mean)
print('The identifier_score_mean', identifier_score_mean)

The semantic valid query in pourcent: 100.0
The keyword_score_mean 0.8733333333333334
The identifier_score_mean 0.3083333333333333


In [ ]:
###### Done in a rush
valid=0
keyword_score_add=0
identifier_score_add=0

p= predictions.copy()
p.pop(28)
p.pop(98)
pred1=p[::2]
pred2=p[1::2]

for pred, gt in zip(pred1, pred2):
  valid_pred, keyword_score, identifier_score = compute_metrics(pred, gt)
  if valid_pred:
    valid+=1
  keyword_score_add+=keyword_score
  identifier_score_add+=identifier_score

valid_pourcent=(valid/len(pred1))*100
keyword_score_mean=keyword_score_add/len(pred1)
identifier_score_mean=identifier_score_add/len(pred1)

print("-----------------------------")

print(valid_pourcent)
print(keyword_score_mean)
print(identifier_score_mean)

Evaluation with Schema of the database in the prompt

In [77]:
base_fewshot_prompt = """
You are an expert in SQL and database querying.
Always return **only** the SQL query, without any explanation, preamble, or additional text.
I give you the database schema wich appears in this form : name_of_the_table : Column_name (Type_of_data), ...| table_2 : ....

here's the database schema : {schema}

Your task is to convert natural language question after <<<>>> into optimized SQL queries. Do not provide any other Sentence.

<<<
Sentence: {sentence} Answer:
>>>
####
Here are some examples:
{examples}
####

""".strip()

In [120]:
import numpy as np

def get_schema(example):
  return db_schema.loc[db_schema['db_id'] == example['db_id'], 'Schema (values (type))'].values

def few_shot_prompt(inp, schema, nb_examples = 5):
  question = inp['question']
  database_id = inp['db_id']
  db= db_schema.loc[db_schema['db_id'] == database_id, 'Schema (values (type))'].values
  db_schema_str = str(db)

  prompt = (
        "You are an expert in SQL and database querying.\n"
        "Always return **only** the SQL query, without any explanation, preamble, or additional text.\n"
        "I give you the database schema which appears in this form:\n"
        "table_name : column_name (data_type), ... | table_2 : ...\n"
        f"Here's the database schema:\n{db_schema_str}\n\n"
        "Your task is to convert natural language question after <<<>>> into optimized SQL queries. Do not provide any other Sentence."
        "<<<Sentence: {sentence} Answer:>>>"
        "Here are some examples:"
    )


  pool_of_examples = np.random.choice(train, nb_examples)
  for example in pool_of_examples:
    prompt += f"Sentence: {example['question']} " + f"Answer:{example['query']}\n"
    prompt += "<<<"
  prompt += "Sentence: " + question + " Answer:"
  return prompt

In [121]:
predictions_4=[]
ground_truths_4=[]

for ex in tqdm(test):
  sch = get_schema(ex)
  prompt = few_shot_prompt(ex, sch, 5)
  answer = generate_2(prompt)
  pred=extract_sql_answer(answer)
  predictions_4.append(pred)
  ground_truths_4.append(ex["SQL_query"])

100%|██████████| 1024/1024 [1:15:16<00:00,  4.41s/it]


In [122]:
for pred,gt in zip(predictions_4, ground_truths_4):
  print(pred)
  print(gt)

SELECT count(*) FROM singer
SELECT count(*) FROM singer
SELECT COUNT(*) FROM singer
SELECT count(*) FROM singer
SELECT Name,  Country,  Age FROM singer ORDER BY Age ASC
SELECT name ,  country ,  age FROM singer ORDER BY age DESC
SELECT name,  country,  age FROM singer ORDER BY age DESC
SELECT name ,  country ,  age FROM singer ORDER BY age DESC
SELECT AVG(Age),  MIN(Age),  MAX(Age) FROM singer WHERE Country  =  'France'
SELECT avg(age) ,  min(age) ,  max(age) FROM singer WHERE country  =  'France'
SELECT AVG(age), MIN(age), MAX(age) FROM singer WHERE country = "France"
SELECT avg(age) ,  min(age) ,  max(age) FROM singer WHERE country  =  'France'
SELECT T1.Song_Name,  T1.Song_release_year FROM singer AS T1 JOIN singer_in_song AS T2 ON T1.Singer_ID  =  T2.Singer_
SELECT song_name ,  song_release_year FROM singer ORDER BY age LIMIT 1
SELECT T1.Name, T1.Song_release_year FROM singer AS T1 JOIN singer_in_song AS T2 ON T1.Singer_ID  =  T2.Singer_ID JOIN song
SELECT song_name ,  song_release

In [123]:
valid=0
keyword_score_add=0
identifier_score_add=0

for pred, gt in zip(predictions_4, ground_truths_4):
  valid_pred, keyword_score, identifier_score = compute_metrics(pred, gt)
  if valid_pred:
    valid+=1
  keyword_score_add+=keyword_score
  identifier_score_add+=identifier_score

valid_pourcent=(valid/len(predictions))*100
keyword_score_mean=keyword_score_add/len(predictions)
identifier_score_mean=identifier_score_add/len(predictions)

print('The semantic valid query in pourcent:', valid_pourcent)
print('The keyword_score_mean', keyword_score_mean)
print('The identifier_score_mean', identifier_score_mean)

The semantic valid query in pourcent: 84.1796875
The keyword_score_mean 0.8281613731320746
The identifier_score_mean 0.34963766431051607


In [72]:
base_fewshot_prompt = """
You are an expert in SQL and database querying.
Always return **only** the SQL query, without any explanation, preamble, or additional text.
I give you the database schema wich appears in this form : name_of_the_table : Column_name (Type_of_data), ...| table_2 : ....

here's the database schema : {schema}
Your task is to convert natural language question after <<<>>> into optimized SQL queries. Do not provide any other questions.
<<<
Sentence: {sentence} Answer:
>>>
"""

In [ ]:
from tqdm import tqdm

predictions_2=[]
ground_truths_2=[]

for ex in tqdm(test):
  prompt = base_fewshot_prompt.format(examples=demo,sentence=ex["question"], schema = db_schema.loc[ex["db_id"], 'Schema (values (type))'])
  answer = generate_2(prompt)
  pred=extract_sql_answer(answer)
  predictions_2.append(pred)
  ground_truths_2.append(ex["SQL_query"])

  0%|          | 0/1024 [00:00<?, ?it/s]/usr/local/lib/python3.10/dist-packages/transformers/generation/configuration_utils.py:628: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.3` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`.
  warnings.warn(
 94%|█████████▍| 965/1024 [53:44<03:28,  3.53s/it]

In [63]:
print(prompt)
print(answers)
print(predictions)
print(ground_truths)


You are an expert in SQL and database querying.
Always return **only** the SQL query, without any explanation, preamble, or additional text.
I give you the database schema wich appears in this form : name_of_the_table : Column_name (Type_of_data), ...| table_2 : ....

here's the database schema : stadium : Stadium_ID (number) , Location (text) , Name (text) , Capacity (number) , Highest (number) , Lowest (number) , Average (number) | singer : Singer_ID (number) , Name (text) , Country (text) , Song_Name (text) , Song_release_year (text) , Age (number) , Is_male (others) | concert : concert_ID (number) , concert_Name (text) , Theme (text) , Stadium_ID (text) , Year (text) | singer_in_concert : concert_ID (number) , Singer_ID (text)
Your task is to convert natural language question after <<<>>> into optimized SQL queries. Do not provide any other questions.
<<<
Sentence: What are  the different countries with singers above age 20? Answer:
>>>

['SELECT COUNT(*) FROM singer;\n<<<\nSenten

In [74]:
for pred,gt in zip(predictions_2, ground_truths_2):
  print(pred)
  print(gt)

SELECT COUNT(*) FROM singer
SELECT count(*) FROM singer
SELECT COUNT(*) FROM singer
SELECT count(*) FROM singer
SELECT name, country, age FROM singer ORDER BY age DESC
SELECT name ,  country ,  age FROM singer ORDER BY age DESC
SELECT singer.Name, singer.Country, singer.Age FROM singer ORDER BY singer.Age DESC
SELECT name ,  country ,  age FROM singer ORDER BY age DESC
SELECT AVG(Age), MIN(Age), MAX(Age) FROM singer WHERE Country = 'France'
SELECT avg(age) ,  min(age) ,  max(age) FROM singer WHERE country  =  'France'
SELECT AVG(Age), MIN(Age), MAX(Age) FROM singer WHERE Country = 'French'
SELECT avg(age) ,  min(age) ,  max(age) FROM singer WHERE country  =  'France'
SELECT DISTINCT(s.Song_Name), s.Song_release_year FROM singer AS s JOIN singer_in_concert AS si ON s.Singer_ID = si.Singer_ID JOIN concert AS c
SELECT song_name ,  song_release_year FROM singer ORDER BY age LIMIT 1
SELECT DISTINCT(s1.Name), s1.Song_release_year FROM singer AS s1 JOIN singer_in_concert AS si1 ON s1.Singer_I

In [76]:
valid=0
keyword_score_add=0
identifier_score_add=0

for pred, gt in zip(predictions_2, ground_truths_2):
  valid_pred, keyword_score, identifier_score = compute_metrics(pred, gt)
  if valid_pred:
    valid+=1
  keyword_score_add+=keyword_score
  identifier_score_add+=identifier_score

valid_pourcent=(valid/len(predictions))*100
keyword_score_mean=keyword_score_add/len(predictions)
identifier_score_mean=identifier_score_add/len(predictions)

print('The semantic valid query in pourcent:', valid_pourcent)
print('The keyword_score_mean', keyword_score_mean)
print('The identifier_score_mean', identifier_score_mean)

The semantic valid query in pourcent: 91.50390625
The keyword_score_mean 0.7831614565989554
The identifier_score_mean 0.3644798642113096
